8297389: resexhausted003 fails with assert(!thread->owns_locks()) failed: must release all locks when leaving VM

TobiHartmann opened the pull request:
`Method::build_profiling_method_data` acquires the `MethodData_lock` when initializing `Method::_method_data` to prevent multiple allocations by different threads. The problem is that when metaspace allocation fails and `JvmtiExport::should_post_resource_exhausted()` is set, we assert during the `ThreadToNativeFromVM` transition in JVMTI code.

Since concurrent initialization is a rare event, I suggest to get rid of the lock and perform the initialization with a `cmpxchg`, similar to how method counters are initialized:
https://github.com/openjdk/jdk/blob/f4b5065c37e86f4b2ca26da6ce678febe4a52950/src/hotspot/share/oops/method.cpp#L644-L646

Since [current code](https://github.com/openjdk/jdk/blob/f4b5065c37e86f4b2ca26da6ce678febe4a52950/src/hotspot/share/oops/method.inline.hpp#L41-L46) in `Method::set_method_data` uses a `Atomic::release_store`, I added a `OrderAccess::release()`.

Thanks,
Tobias

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297389](https://bugs.openjdk.org/browse/JDK-8297389): resexhausted003 fails with assert(!thread->owns_locks()) failed: must release all locks when leaving VM


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**) ⚠️ Review applies to [7e94c2b2](https://git.openjdk.org/jdk/pull/11316/files/7e94c2b28b04cb5477eddd5f66126d05ead70119)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11316/head:pull/11316` \
`$ git checkout pull/11316`

Update a local copy of the PR: \
`$ git checkout pull/11316` \
`$ git pull https://git.openjdk.org/jdk pull/11316/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11316`

View PR using the GUI difftool: \
`$ git pr show -t 11316`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11316.diff">https://git.openjdk.org/jdk/pull/11316.diff</a>

</details>

TobiHartmann commented:
/label hotspot

dholmes-ora commented on src/hotspot/share/oops/method.cpp:
```diff
@@ -590,26 +590,29 @@ void Method::build_profiling_method_data(const methodHandle& method, TRAPS) {
+  // The store into method must be released. On platforms without
+  // total store order (TSO) the reference may become visible before
+  // the initialization of data otherwise.
+  OrderAccess::release();
```
A release here is not necessary as `Atomic::replace_if_null` will by default have `memory_order_conservative` which maintains the full bi-directional fence of the internal `cmpxchg`.

dholmes-ora commented on src/hotspot/share/oops/method.cpp:
```diff
@@ -590,26 +590,29 @@ void Method::build_profiling_method_data(const methodHandle& method, TRAPS) {
-      return;   // return the exception (which is cleared)
-    }
+  ClassLoaderData* loader_data = method->method_holder()->class_loader_data();
+  MethodData* method_data = MethodData::allocate(loader_data, method, THREAD);
```
So the downside of lock-free here is that we have to pre-allocate and then later free. How expensive is that? How likely are we to get multiple threads attempting this at the same time? We might trigger a resource-exhausted event unnecessarily due to the temporary use of metaspace.

dholmes-ora reviewed (changes requested):
If `MethodData::allocate` cannot be called whilst holding a lock then perhaps we can assert that in there?

I think there is a broader problem here that metaspace allocation can occur in numerous places and any failure could post the resource-exhausted event and potentially lead to problems like this.

This fix side-steps one problematic call-site, but going lock-free has its own concerns.

TobiHartmann commented:
Thanks for the review, David!

I added an assert to `MethodData::allocate` and fixed the ciReplay code which only ever gets executed by a single thread by removing the lock. I also removed the `OrderAccess::release()`.

Regarding the overhead of going lock-free: I temporarily added code that counts the number of times that multiple threads attempt initialization and asserts if it's more that 50x. This triggers in 25 out of 203.772 runs (tests from tier1 to tier3). The average size of these allocations is around 88 words (704 bytes). I therefore think it's fine to avoid a lock in this case. Also, we already use a lock-free mechanism for `method::build_method_counters`.

What do you think?

reinrich reviewed (approved):
Thanks for taking care of the issue. The fix looks ok to me. Seems like you've removed all uses of `MethodData_lock` so you could remove the declaration and definition too.

Thanks, Richard.

TobiHartmann commented:
Thanks for the review, Richard. Good catch, I removed the `MethodData_lock`.

dholmes-ora commented:
> I therefore think it's fine to avoid a lock in this case. Also, we already use a lock-free mechanism for `Method::build_method_counters`.

`Method::build_method_counters` doesn't appear to be lock-free, in the sense there are no atomic operations; rather concurrency just doesn't seem to be a concern with that code. ??

But as long as the overhead is low and interference unlikely, then we can see how this works out in practice. Thanks.

TobiHartmann commented:
> I'm not sure if it's guaranteed that replay runs single-threaded. I haven't looked into the details, but I think ReplayInline can run in any compiler thread after startup, and then of course there is JDK-8254110.

But ReplayInline does not load ciMethodData, i.e., does not call `process_ciMethodData`, right?

The only way this code can get executed is through `JNI_CreateJavaVM_inner -> ciReplay::replay -> ciReplay::replay_impl -> process-> process_command -> process_ciMethodData` and that only ever happens single-threaded.

Of course, if we are ever going to implement [JDK-8254110](https://bugs.openjdk.org/browse/JDK-8254110), we need to revisit that code, but in that case the assert that I added will trigger.

What do you think?

> Method::build_method_counters doesn't appear to be lock-free, in the sense there are no atomic operations; rather concurrency just doesn't seem to be a concern with that code. ??

It uses the exact same mechanism that I now added to `Method::build_profiling_method_data`, including atomic operations to initialize `_method_counters`:
https://github.com/openjdk/jdk/blob/f4b5065c37e86f4b2ca26da6ce678febe4a52950/src/hotspot/share/oops/method.cpp#L651-L654

And I verified that multiple threads attempt to initialize the counters by adding verification code to:
https://github.com/openjdk/jdk/blob/f4b5065c37e86f4b2ca26da6ce678febe4a52950/src/hotspot/share/oops/method.cpp#L644-L646

dholmes-ora commented:
> It uses the exact same mechanism that I now added to Method::build_profiling_method_data, including atomic operations to initialize _method_counters

Sorry I somehow missed the `init_method_counters` logic. :(

dean-long commented:
> Of course, if we are ever going to implement [JDK-8254110](https://bugs.openjdk.org/browse/JDK-8254110), we need to revisit that code, but in that case the assert that I added will trigger.
> 
> What do you think?

OK, please add a TODO in JDK-8254110 explaining this so it doesn't get lost.

TobiHartmann commented:
> OK, please add a TODO in JDK-8254110 explaining this so it doesn't get lost.

Thanks, done.

dholmes-ora reviewed (approved):
Changes seem okay but I'm not that familiar with the actual code here.

Thanks.

TobiHartmann commented:
Thanks, David!

TobiHartmann commented:
Thanks, Dean!

TobiHartmann commented:
/integrate
